### ISSUE TITLE
Acquia + BLT + Pipelines + CDE Deployments Failing Due to memcache and concurrency

### ISSUE THREAD

**Issue by aweingarten** [acquia/blt#3352]
**DISCLAIMER: THIS IS NOT A BLT PROBLEM** 
It is an Acquia Cloud, CD and Acquia Search problem.  However, BLT users will get burned in different ways. Documenting it here may save other people pain and suffering.  

**Source of Problem**
Acquia released updates to memcache AND pipelines the week of 1/13 which caused massive regressions in our CI/CD pipeline.  

**Memcache**
We are using config split.  We have never enabled DBLOG in prod.  We enable it via config split in lower envs.  After the hosting release,  when we tried to pull down DBs, we began getting errors complaining about the DBLOG table not existing.  It seems that, immediately after the DB copy. memcache still had a cached service container from before the copy and thought that the DBLOG module was enabled before it actually was.  It then tried to write to a non-existent table and failed.   

We were able to solve this by adding a drush CR to the db cloud hooks before config import operations.

**Concurrency In Pipelines**
We are using github -> pipelines -> CDE as our primary development pipeline.  When a PR passes a build it enables an ODE and deploys code to it.  It also copies a DB over to that ODE.  We also use BLT's stock cloud hooks to do`updb` and config imports/ 

This process used to work without concurrency issues.  With the recent code deployments the two cloud tasks are running in parallel and trampling each other.  The cloudhooks for a DB sync may error because of missing composer files since the file deployment may be in progress, it could also fail because another drush CR is in progress.  Sometimes they may just pass.  So far the only solution we have found is to add a **3 minute** sleep to our DB hooks to force a wait for the code deployment to complete.  Its a nasty solution.  

**Acquia Search Config Issues**
The last issue we encountered was that BLT's config validation functionality was failing on the cloud hooks on the CDE's.  The additional CR we added seems to have gotten Acquia Search to fire a hook which pulls in the possible cores and stores them in active config.  Since these will always change for CDE's it means that the CMI will never validate.  To work around we were forced to add config-ignore and ignore the acquia-search possible cores.  This effectively converts them from config to state variables as they should have been from the beginning.   NOTE this issue has been documented and known since at least **5/17. See[ D.O Issue Here](https://www.drupal.org/project/acquia_connector/issues/2864697)**

**Post Mortem**
It's pretty clear that Acquia released a massive hosting release without testing the BLT -> Pipelines -> CDE integrations, which is a real shame.  My guess is that they tested pipelines builds and maybe that the containers were created, but did not understand the integration points of blt/cloud-hooks with the system.    

**TLDR Solution**
- Use config-ignore on Acquia search
- 1 CR before DB syncs in acquia cloud update hooks
- Add a 3 minute sleep to db post copy hook on CDE

**Comment by aweingarten** [comment]
@danepowell, is there anything we should do to make sure that others know this is still an issue?

**Comment by danepowell** [comment]
Please report these issues to your account manager or open support tickets, and make sure to clearly specify which product has the issue (Pipelines, Search, etc...). That will ensure that they get triaged and tracked appropriately.

For the Pipelines issue, I wonder if the recent change to the Pipelines yml config might help you. Previously, there was a race condition caused by how some projects deployed to CDEs, which is fixed by moving the deploy script to the post-deploy event: https://github.com/acquia/blt/pull/3374

**Comment by aweingarten** [comment]
Hey Dane, we actually have already escalated these issues with the AM team. We also were the source of the changes to the pipelines files you mentioned. The concurrency issues have not been fixed for the past 8 weeks.  The lack of a timely resolution, starts to put this into the scope of BLT to work around or band-aid.  

Ditto for memcache and search issues as the problems we encountered would likely impact other BLT users.

**Comment by lcatlett** [comment]
@aweingarten - thank you for following up with this in the BLT queue! For those watching this issue or who have come across it in search results, this bug was reviewed by the Acquia product team and was put into the product backlog as a high priority item to fix due to the critical issues you reported and the fact that it impacted multiple products. I can't speak to a specific release date on behalf of the pipelines and cloud teams, however, Acquia engineers are working alongside subject matter experts in caching, CI, deployment, and d8 best practices to resolve this and related issues discovered during our initial debugging. I have spoken to your account manager and confirmed that you will be updated as soon as we have more information and recommended next steps.

**Comment by danepowell** [comment]
> The lack of a timely resolution, starts to put this into the scope of BLT to work around or band-aid.

While I know that folks have used BLT as this sort of stop-gap solution in the past, that is explicitly no longer part of BLT's mission. As Lindsey mentioned, there's an explicit process for triaging and prioritizing product issues going forward, and you will likely see elements of BLT (such as memcache settings) getting de-scoped in the near future as they are incorporated into other products.

**Comment by aweingarten** [comment]
@danepowell this is a pretty major change for the BLT project to communicate to the community inside a closed issue.  

Would it be possible for you to open an issue to expand on this and reset people's expectations?  What is the purpose of BLT going forward?  IF the fundamental goals of BLT are changing it would be great to know how we as members of the community should engage with you going forward and what we should expect from the BLT team.   What are valid submissions or issues?

**Comment by danepowell** [comment]
Nothing fundamental is changing. Being a band-aid has never been part of BLT's mission, as reflected in our README and project homepage: https://github.com/acquia/blt#philosophy-and-purpose

We will be revising the issue template and contribution guidelines shortly to add additional guidance on how folks can get support, along the lines of what Lindsey and I have outlined here.
